[AVS] Convert to typescript

acrollet opened the pull request:
## Are you removing, renaming or moving a folder in this PR?
- [x] No, I'm not changing any folders (skip to TeamSites and delete the rest of this section)
### :warning: TeamSites :warning:

Did you change site-wide styles, platform utilities or other infrastructure?
- [x] No

## Summary

Migrates AVS app to TypeScript.

## Cross-team Impact: Risks & Mitigations

This PR includes several changes that affect shared configuration files. Below is an assessment of each risk area and the mitigations applied.

### Prettier config & version (REVERTED — no risk)

The original PR bumped prettier from `^1.14.3` to `^2.8.8` and added `arrowParens: "avoid"` and `bracketSameLine` globally. This would have caused noisy reformatting diffs on every JS file touched by lint-staged.

**Mitigation:** All prettier changes have been reverted. `.prettierrc` and the prettier version in `package.json` are identical to `main`. Prettier 1.x already supports TypeScript natively (since v1.4), so no config changes are needed.

### Mocha / Babel register config (global, low risk)

`config/mocha.json` replaces `@babel/register` with a custom `config/babel-register.js` that adds `@babel/preset-typescript` and `.ts`/`.tsx` extensions.

- **`@babel/preset-typescript`** strips type annotations. For plain JS files there are no annotations to strip, so this is a no-op for non-TS apps.
- **Extension list** adds `.ts`/`.tsx` alongside existing JS extensions. Does not affect how `.js`/`.jsx` files are processed.
- **`cache: false`** was originally set but has been removed, so Babel's default caching is preserved and other apps' test performance is unaffected.

### ESLint TypeScript overrides (global, low risk)

New override blocks in `.eslintrc.js` and `.eslintrc.changed.js` are scoped to `**/*.ts` and `**/*.tsx` file patterns. They will not activate for JS/JSX files.

The `.eslintrc.js` override includes `project: ['./src/applications/*/tsconfig.json']`. This glob only resolves when ESLint processes `.ts`/`.tsx` files. Apps without a `tsconfig.json` won't have `.ts` files, so a non-matching glob won't cause errors.

### CI typecheck job (global, low risk)

A new `typecheck` job was added to both CI workflows:

- **`continuous-integration.yml` (main):** Runs with `continue-on-error: true` and is **not** in the build job's `needs` list. Typecheck failures on main are informational only and will never block builds.
- **`pull-request.yml` (PRs):** Runs typecheck scoped to apps that have a `tsconfig.json`. Can be made a required check via branch protection for TS-enabled apps only.

Both workflows discover tsconfigs dynamically via `find src/applications -name 'tsconfig.json'`, so they only check apps that have opted into TypeScript.

### @babel/traverse resolution (global, low risk)

Added `"**/@babel/traverse": "^7.25.9"` to `package.json` `resolutions` to fix a pre-existing version fragmentation issue (nested 7.15.0 under `@babel/preset-env` missing `visitors.environmentVisitor`). This deduplicates all copies to the root version already in use, reducing `yarn.lock` size.

### New devDependencies

- `@babel/preset-typescript@^7.27.1` — needed for Babel to handle `.ts`/`.tsx` in tests
- `@typescript-eslint/eslint-plugin@^5.62.0` and `@typescript-eslint/parser@^5.62.0` — needed for ESLint TS overrides
- `@types/react@^17.0.0` — TypeScript type definitions for React (downgraded from 18 to match the React version in use)

acrollet committed:
Migrate AVS application to TypeScript and update dependencies

acrollet committed:
Remove AVS application components and related files

acrollet committed:
Refactor AVS application types and utilities: add medication types, restructure constants, and implement phone number normalization

acrollet committed:
feat: Add new AVS components and types for enhanced health information display

acrollet committed:
Refactor test imports to include file extensions

- Updated import statements in unit tests to include the appropriate file extensions (.tsx, .ts) for components, loaders, and utilities.
- Modified the yarn.lock file to update dependencies and their versions.

acrollet committed:
Merge branch 'main' into feature/avs-ts

acrollet committed:
move platform types to platform dir

acrollet committed:
Merge branch 'feature/avs-ts' of https://github.com/department-of-veterans-affairs/vets-website into feature/avs-ts

acrollet committed:
Refactor error handling in AvsErrorElement and ErrorBoundary components

acrollet committed:
Merge branch 'main' into feature/avs-ts

mdewey commented on src/applications/avs/types/declarations/va-components.d.ts:
```diff
@@ -0,0 +1,43 @@
+// Type declarations for VA web components used as JSX elements
+
+declare namespace JSX {
```
SUGGESTION: is there a better place for these?

mdewey commented on src/applications/avs/types/index.ts:
```diff
@@ -0,0 +1,297 @@
+// Common types for the AVS application
```
SUGGESTION: break this into a few files? Or are big type files a TS thing?

mjknight50 replied:
No, a "big file" isn't a TS thing.  
Often what I do is create a /models folder and subfolders for types.

There isn't anything inherently wrong with this current one, but it will be unmanageable as it grows.

mjknight50 replied:
<img width="550" height="336" alt="Screenshot 2025-09-19 at 11 34 51 AM" src="https://github.com/user-attachments/assets/2068bff5-7ab1-4a8d-bc85-25ba25669af9" />

mdewey commented on src/applications/avs/tests/components/AvsPageHeader.unit.spec.jsx:
```diff
@@ -2,7 +2,7 @@ import React from 'react';
 import { render } from '@testing-library/react';
 
-import AvsPageHeader from '../../components/AvsPageHeader';
+import AvsPageHeader from '../../components/AvsPageHeader.tsx';
```
QUESTION: Do we need to .tsx the tests as well?

mdewey reviewed (dismissed):
Looks good, I ran it both as it self and as part of the whole website and things worked

QUESTION: DOes the API response get a type? I am not seeing it but I could be just not seeing it

acrollet committed:
Merge branch 'main' into feature/avs-ts

acrollet committed:
Add AvsApiResponse type and update avsLoader to use typed response

acrollet committed:
Refactor: Update VaTelephone component usage and remove outdated type definitions

acrollet committed:
feat(avs): migrate from custom va-component types to component library

- Replace all va-* web components with React components from @department-of-veterans-affairs/component-library
- Update va-telephone, va-alert, va-loading-indicator, va-accordion, va-accordion-item, va-additional-info, and va-icon
- Remove custom type definitions file (va-components.d.ts)
- Convert kebab-case attributes to camelCase props (not-clickable → notClickable)
- Remove deprecated uswds props where not supported
- Improve type safety by using official component library types

Components updated:
- Avs.tsx: VaLoadingIndicator, VaAccordion, VaAccordionItem
- ErrorBoundary.tsx: VaAlert
- YourAppointment.tsx: VaIcon
- MedicationTerms.tsx: VaAdditionalInfo
- MoreInformation.tsx: VaTelephone
- YourHealthInformation.tsx: VaTelephone

acrollet committed:
refactor(avs): break up types file into logical modules

- Extract domain types (medical data, AVS core types) to types/domain.ts
- Extract API response types to types/api.ts
- Extract UI component props to types/ui.ts
- Extract common utility types to types/common.ts
- Convert index.ts to barrel export for backward compatibility
- Maintains all existing imports while improving organization and maintainability

acrollet committed:
restore fancy apostrophe

acrollet commented:
> QUESTION: DOes the API response get a type? I am not seeing it but I could be just not seeing it

added!

acrollet committed:
fix(avs): update root URL redirection logic to handle multiple valid paths

acrollet committed:
Fix logic

acrollet committed:
Merge branch 'main' into feature/avs-ts

acrollet committed:
Change feature toggle approach

mjknight50 commented on src/applications/avs/components/AvsErrorElement.tsx:
```diff
@@ -0,0 +1,44 @@
+  badRequest: 'bad_request',
+} as const;
+
+interface AsyncError {
```
This could be an enum

export enum ErrorTypes {
unauthorized = 'unauthorized',
  notFound = 'not_found',
  badRequest = 'bad_request',
}

And I would move it to a common location like /data/enums

mjknight50 replied:
Move to a common location /data/interfaces

mjknight50 commented on src/applications/avs/utils/medications/index.ts:
```diff
@@ -0,0 +1,83 @@
+};
+
+/**
+ * Adds medication source to all medications in the array
```
Nice comments!

acrollet committed:
refactor(avs): improve error handling using React Router best practices

- Replace custom AsyncError interface with isRouteErrorResponse type guard
- Add type-safe error handling for React Router ErrorResponse objects
- Maintain support for VA.gov API error format and standard HTTP responses
- Add comprehensive unit tests covering all error scenarios (10 test cases)
- Improve code maintainability and future-proofing with React Router standards

All tests passing: 135 passing, 7 pending

acrollet committed:
refactor(avs): extract error types to shared enum

- Create ErrorTypes enum in src/applications/avs/types/enums.ts
- Export enum from main types index for easy importing
- Replace inline error type object with reusable enum
- Improve type safety and consistency across the application
- Enable better IDE support and refactoring capabilities

All tests passing: 135 passing, 7 pending

acrollet committed:
Add TypeScript support to precommit hooks

- Update lint-staged in package.json to include .ts and .tsx files
- Install @typescript-eslint/parser and @typescript-eslint/eslint-plugin v5.62.0 compatible with Node.js v14
- Add TypeScript overrides to both .eslintrc.js and .eslintrc.changed.js
- Configure TypeScript parser with proper parserOptions and project paths
- Disable conflicting base ESLint rules and enable TypeScript-specific rules
- Add warnOnUnsupportedTypeScriptVersion: false to suppress version warnings
- Fix AvsErrorElement.tsx to use compatible syntax with older Prettier version
- Ensure precommit hooks now validate TypeScript files alongside JavaScript files

acrollet committed:
feat: upgrade prettier and fix all ESLint issues in AVS directory

🚀 Major improvements:
- Upgraded Prettier from 1.14.3 to 2.8.8 for TypeScript support
- Updated .prettierrc to handle TypeScript files with proper parser
- Fixed all 30+ ESLint errors in AVS TypeScript files

🔧 ESLint fixes:
- Replaced @ts-ignore with proper eslint-disable comments
- Removed unused imports (LabResult, ClinicalReminder)
- Fixed banned type {} → Record<string, never>
- Fixed React children prop usage
- Formatted all TypeScript files with new Prettier

✅ Results:
- All TypeScript 'import type' syntax now properly supported
- Zero ESLint errors remaining in AVS directory
- Improved code quality and type safety
- Better developer experience with modern tooling

acrollet committed:
fix: update CI workflows to lint TypeScript files

🔧 CI Updates:
- Updated filter-files-changed.js to include .ts and .tsx files
- Added --ext .ts --ext .tsx to all ESLint commands in CI/PR workflows
- Ensures TypeScript files in AVS directory are properly linted

📝 Changes:
- continuous-integration.yml: Added TS extensions to main linting job
- pull-request.yml: Added TS extensions to both full and incremental linting
- filter-files-changed.js: Updated regex to match TypeScript files

This ensures our new TypeScript files in the AVS directory will be
properly linted in CI/CD pipelines.

acrollet committed:
Merge branch 'main' into feature/avs-ts

acrollet committed:
Update prettier config to match js config

acrollet committed:
Merge branch 'feature/avs-ts' of https://github.com/department-of-veterans-affairs/vets-website into feature/avs-ts

acrollet committed:
lint filter-files-changed.js

acrollet committed:
Merge branch 'main' into feature/avs-ts

acrollet committed:
Add typecheck command

acrollet committed:
Add TypeScript typecheck job to CI workflow and create output parser

acrollet committed:
Update ts lint config

acrollet committed:
TS linting improvements

acrollet committed:
Merge branch 'main' into feature/avs-ts

acrollet committed:
chore: update ESLint configuration for improved module resolution

acrollet committed:
Add ts migration guide

acrollet committed:
Merge branch 'main' into feature/avs-ts

acrollet committed:
refactor: replace Date mocking with sinon fake timers in unit tests

acrollet committed:
Merge branch 'feature/avs-ts' of https://github.com/department-of-veterans-affairs/vets-website into feature/avs-ts

acrollet committed:
Merge branch 'main' into feature/avs-ts

mjknight50 commented on src/applications/ask-va/tests/components/Announcements.unit.spec.jsx:
```diff
@@ -40,19 +40,12 @@ const generateStore = (initialState = {}) => {
 describe('<Announcements />', () => {
   const mockDate = new Date('2024-03-20T12:00:00Z');
-  let originalDate;
+  let clock;
```
I don't how ESLINT is setup, but I often do  const over let
https://eslint.org/docs/latest/rules/prefer-const

Which would change this all:

const clock = sinon.useFakeTimers({ now: mockDate.getTime(), toFake: ['Date'] });
  const apiRequestStub = sinon.stub(apiUtils, 'apiRequest').resolves(mockAnnouncementsResponse);
  const constantsStub = sinon.stub(constants, 'mockTestingFlagforAPI');

mjknight50 commented on src/applications/avs/types/api.ts:
```diff
@@ -0,0 +1,23 @@
+import type { AvsData } from './domain';
+
+// API Response types
+export interface ApiResponseData<T = unknown> {
```
Is this really unknown?
The main difference between unknown and any is that unknown is much less permissive.

It feels like you could define this much better.  
Generally I try to prevent ANY or UNKNOWN in Typescript because then it might as well just be JS garbage.

mjknight50 commented on src/applications/avs/utils/constants.ts:
```diff
@@ -0,0 +1,78 @@
+// Type definitions for constants
```
I could be wrong, but I never use const files like this.  I use enums for this purpose.  It sounds like the TS creators wouldn't use enums now, but I personally like them:
https://www.youtube.com/watch?v=vBJF0cJ_3G0&t=1012s

mjknight50 commented on src/applications/avs/utils/index.ts:
```diff
@@ -2,8 +2,9 @@ import { parse } from 'date-fns';
+import type { AvsData } from '../types';
 
-const parseProblemDateTime = dateString => {
+export const parseProblemDateTime = (dateString: string): Date | string => {
```
There's a lot of date/time stuff here in index.  Should that be moved out to its own utils/datetime.ts file?

acrollet committed:
Merge branch 'main' into feature/avs-ts

acrollet committed:
Merge branch 'main' into feature/avs-ts

acrollet committed:
Remove unnecessary unknown types from AVS application

- Replace generic ApiResponse<T = unknown> with concrete AvsApiResponse type
- Remove index signatures with unknown from Order, AvsData, AppState, and AvsState interfaces
- Keep legitimate unknown uses in generic UI components (ListBlockProps, ItemsBlockProps)
- Improves type safety without breaking changes - API extra fields still handled correctly

Addresses code review feedback about unknown types defeating TypeScript's purpose.

acrollet committed:
Refactor: Extract date/time utilities to separate file

- Move all date/time parsing and formatting functions to utils/datetime.ts
- Keep general utility functions (fieldHasValue, allArraysEmpty, allFieldsEmpty) in utils/index.ts
- Re-export datetime utilities from index.ts for backward compatibility
- No breaking changes - all existing imports continue to work

Improves code organization and separation of concerns.

acrollet committed:
Add comprehensive JSDoc documentation to domain types

- Add interface-level JSDoc comments explaining purpose of each type
- Add property-level JSDoc comments for all fields
- Document medical terminology, data formats, and field purposes
- Improve code maintainability and developer understanding

All 23 interfaces now have complete documentation including:
- ClinicVisited, AvsMetadata, Medication (with 22 properties documented)
- Health information types (Appointment, Immunization, Problem, Allergy, etc.)
- Lab results with test values and metadata
- Treatment plan types (Order, ClinicalReminder, MedicationChangesSummary)
- Visit data types (VitalSign, Diagnosis, Procedure, ClinicMedication)
- Main AvsData interface with 29 properties documented

mjknight50 reviewed (dismissed):
GTG

acrollet committed:
Merge branch 'main' into feature/avs-ts

acrollet committed:
Merge branch 'main' into feature/avs-ts

acrollet committed:
Merge remote-tracking branch 'origin/main' into feature/avs-ts

# Conflicts:
#	.eslintrc.js
#	.github/workflows/continuous-integration.yml
#	package.json
#	src/applications/avs/containers/Avs.tsx
#	yarn.lock

acrollet committed:
Merge main

acrollet committed:
Restore non-AVS applications to main state

acrollet committed:
Fix ESLint: remove unused 'id' destructuring in Avs component

acrollet committed:
CI: allow TS-related deps and restore yarn.lock threshold to 600

acrollet committed:
CI: scope typecheck to TS-enabled apps, non-blocking on main

- CI (main): typecheck runs with continue-on-error, annotations only, never blocks builds
- PR: scoped typecheck runs only on apps with their own tsconfig.json
- Both workflows use find src/applications -name tsconfig.json for discovery
- Non-TS teams are completely unaffected

acrollet committed:
Merge branch 'main' into feature/avs-ts

acrollet committed:
fix: add @babel/traverse resolution to fix version fragmentation

acrollet committed:
fix(avs): address high-priority TS review findings

- ListBlock: map over filtered listItems instead of unfiltered items
- ItemsBlock: use generics instead of double-cast as unknown as
- RootState: consolidate into single definition in types/redux.ts

acrollet committed:
refactor(avs): replace @ts-ignore with @ts-expect-error, remove unnecessary suppressions

acrollet committed:
refactor(avs): consolidate duplicate AppointmentType into types/domain.ts

acrollet committed:
refactor(avs): remove unused AvsLoaderData type

acrollet committed:
refactor(avs): remove unused AppState interface

acrollet committed:
revert: remove prettier config changes and version bump

Prettier 1.x already supports TypeScript. The config changes
(arrowParens, bracketSameLine) and version bump to 2.x would
cause noisy reformatting diffs across all JS files.

acrollet committed:
fix: remove cache:false from babel-register to avoid slowing other apps' tests

acrollet committed:
fix: resolve duplicate @types/react and add skipLibCheck for AVS typecheck

acrollet committed:
Merge branch 'main' into feature/avs-ts

acrollet committed:
fix: update prettier to version 1.19.1

acrollet committed:
Merge branch 'feature/avs-ts' of https://github.com/department-of-veterans-affairs/vets-website into feature/avs-ts

acrollet committed:
fix: upgrade prettier to 2.8.8 for TypeScript import type support

Add arrowParens and endOfLine to .prettierrc to preserve 1.x formatting behavior.
